active/0021-transfer-files-over-mqtt.md

```diff
@@ -0,0 +1,175 @@
+* device may not know total file size when starting the transfer
+
+### Requirements
+
+* Based on the MQTT protocol. Use MQTT messages to encapsulate file bodies
+* Transport layer TCP, TCP/TLS and QUIC (TCP fallback) multiplexing connections with other service topics​
+* High concurrency (multiple uploads in progress)
+* MQTT Client single/bi-directional transfer (active file upload, passive file download)
+* Fault tolerance: disconnected retransmission (sender side)
+* Fragmentation: (multi parts, fragments)
+* Multi-channel: multiple TCP connections, QUIC multi-streams
+* Cancellation: Both parties can pause or cancel the transfer, and an event notification is required when the transfer is complete. (How to notify? What is the target of the notification?)
+* Minimal code changes on Client side, SDK support available. (low code/no code)
+* Minimize data retransmission, and reduce network pressure and bandwidth usage.
+* RestAPI interface Query transfer status
+* EMQX local incomplete file cache
+* Client API should be easy to use: easy to migrate from existing applications (ftp, nfs, s3)
+* File body Encryption and decryption (with external key management)?
+
+### Out of scope
+
+* Non-reliable transmission is not supported
+* No support for one-to-many transfers
+* No distributed cache storage
+* Broker does not aware of interrupted transfer, It will be client-side implementation
+
+### Limitations
+
+* File transfer is PUSH mechanism but not support POLL, the receiving side cannot request to initiate the transfer
+decisions of the size of a slice cannot be as natural as streaming.
+* Due to network bandwidth limitations, it may happen that multiple files are transferred in parallel but all fail to complete the transfer in the desired time.
+* Do not get too hung up on the length of the topic and do not take advantage of "topic alias" (MQTT 5.0)
+* Performance may be limited by disk IO (network in speed > disk out speed)
+* Overwrite existing file SEG
+* Linux filename length 255 bytes. path_max 4096
+* EMQX terminates file transfers, file message MQTT forwarding (as proxy) is not required.
+* QUIC in EMQX does not support HTTP/3 protocol, old client supports HTTP upload, how to smooth migration?
+
+### Protocol design
+
+* Maximum bandwidth utilization, high signal-to-noise ratio (DATA/ MQTT HEADER)
+* Asynchronous stateless non-continuous transfer, sliceable, messy (retransmission after disconnection) (parallel transfer of multiple files)
+* Fault tolerance.
+  * Incomplete packets Disconnected.
+  * Transmission can be paused or resumed for a long time
+  * Interactive completions required???
+* Local caching of file contents
+* Utilizes only MQTT topic and payload (Payload Format Indicator?). No header changes.
+* TCP multi-connection support (using Same Client ID)
+* User could manipulate file names, and paths. (blacklist/check file names/path, or relative path only)
```

Comment:
Suggestion: 
* First publish: `$file/sha256sum/init` to initiate the transfer, and the payload can be a json body of file metadata (whatever the user wants to put in)
* Following messages: `$file/sha256sum/0` as topic name, payload for file content.

Reply:
> Suggestion:
> 
> * First publish: `$file/sha256sum/init` to initiate the transfer, and the payload can be a json body of file metadata (whatever the user wants to put in)
> * Following messages: `$file/sha256sum/<byte-offset>` as topic name, payload for file content.

We will actually need a protocol for this init message in order to communicate the total size of the file etc.

Reply:
As far as I understand there are use cases when client would not know file size in advance, like rolling log file.

Reply:
How would broker deal with file metadata? Is it going to be a separate .json with the same name?